internal/route/router.go

```diff
@@ -143,32 +155,48 @@ func (r *SimpleRouter) Route(
 		HTTPReponseCode = strconv.Itoa(response.StatusCode)
 	}
 
-	isJSONRPCError := body != nil && body.Error != nil
-	if err != nil || isJSONRPCError {
-		JSONRPCResponseCode := ""
-
-		if isJSONRPCError {
-			zap.L().Warn("Encountered error in upstream JSONRPC response.", zap.Any("request", requestBody),
-				zap.Any("error", body.Error), zap.String("client", util.GetClientFromContext(ctx)), zap.String("upstreamID", id))
-
-			JSONRPCResponseCode = strconv.Itoa(body.Error.Code)
-		}
-
+	if err != nil {
 		metrics.UpstreamRPCRequestErrorsTotal.WithLabelValues(
 			util.GetClientFromContext(ctx),
 			id,
 			configToRoute.HTTPURL,
-			requestBody.Method,
+			requestBody.GetMethod(),
 			HTTPReponseCode,
-			JSONRPCResponseCode,
+			"",
 		).Inc()
 	}
 
+	if body != nil {
+		// To help correlate request IDs to responses.
+		// It's the responsibility of the client to provide unique IDs.
+		reqIDToRequestMap := make(map[int]jsonrpc.SingleRequestBody)
+		for _, req := range requestBody.GetSubRequests() {
+			reqIDToRequestMap[int(req.ID)] = req
+		}
```

Comment:
A couple thoughts / questions on this mapping:
1. Is it safe to assume that the responses will always be in the same order as the requests? If it is, then we should be able to match a response to the corresponding request by index only, without having to rely on additional IDs.
2. If we do need the IDs, should we add some lightweight error handling if they're not unique? It seems reasonable to make it the client's responsibility, but we might want to inform them if it happens.

Reply:
The [JSON RPC spec](https://www.jsonrpc.org/specification#batch) states that batches can be returned in any order:
```
The Response objects being returned from a batch call MAY be returned in any order within the Array. The Client SHOULD match contexts between the set of Request objects and the resulting set of Response objects based on the id member within each Object.
```

However on that page I also read:
```
A Response object SHOULD exist for each Request object, except that there SHOULD NOT be any Response objects for notifications.
```
Notifications are `Request object without an "id" member` and also `The Server MUST NOT reply to a Notification, including those that are within a batch request.` so I think we're good for handling notifications.

Reply:
Regarding error handling if IDs aren't unique:
* I think our gateway should be more of a "pass-through" and leave validation up to the backend services. 
* I can't think of a good way to emit a warning back to the client. The request should be able to get served properly by backends.
* Right now we're only using the IDs for our own metrics, so it's not _super_ critical if we don't have unique IDs.